### PR TITLE
[Tech] Refactor game managers

### PR DESCRIPTION
**Pull request by BrettCleary**
## Summary
This refactors all store managers (hyperplay, gog, sideload, legendary) to use the same module level interface
- previously hyperplay and sideload has functional module exports but gog and legendary used singletons and OOP inheritance
- this decouples and abstracts away any store that could be integrated with HP, making it easier to test, work with, and extensible to more stores in the future

Sideload Game info has been added into `GameInfo` which makes sense because the frontend shouldn't be concerned with implementation level backend details like what store/runner the game is using

HyperPlay store methods have been implemented and improved

Overlay injection occurs using process names now so bootstrap exes and processes that perform checks before launching the actual game process can now be launched without losing the in game overlay injection

`Installing` reads as `Downloading` now on game card and there is an `Extracting` state

Import game takes a path to the exe instead of the parent folder

Fixes a bug with the default install path setting

## Not Implemented Yet
`moveInstall` and `changeInstallPath` for the hyperplay store manager

## Tested

On Windows and SteamDeck
- Install, Launch, Uninstall, Update, Refresh, Import with HyperPlay, GOG, and Epic games store managers
- Add Game, Launch, Remove game for sideload store manager

### CODE REVIEW

**Comment by hellowodl** [comment]
This branch seems to have fixed HP games not launching through proton on Linux.
GOG proton games seem to work as well.

I currently don't have any games on Epic, so I haven't tested those. Though basic functionality works great(er than before) on Linux.

File: src/backend/main.ts [hunk 1]
```diff
@@ -159,19 +147,39 @@ import {
...
-} from './hyperplay/library'
+import { addGameToLibrary } from './storeManagers/hyperplay/library'
 
-import {
-  importGame,
-  isHpGameAvailable,
-  isHpGameNative,
-  stopHpGame
-} from './hyperplay/games'
+import * as HyperPlayGameManager from './storeManagers/hyperplay/games'
```
Comment (hellowodl) [inline src/backend/main.ts:152]: Would it be an idea to move the map composition to ./storeManagers/index.ts, in order to de-clutter main.ts a wee bit?

**Review by hellowodl — APPROVED** [review]
Massive improvement, I've tested this with GOG & HyperPlay on Linux.
Nonetheless it might be worth it for someone to test it with Epic games on Linux as well.

Whilst the PR has my approval, it needs to be tested on MacOS as well.

**Comment by hellowodl** [comment]
> @hellowodl since this is a big change can you create a list similar with the one I posted above and test on Linux if you have the time?

Sure thing

**Comment by hellowodl** [comment]
```
(node:47213) UnhandledPromiseRejectionWarning: ReferenceError: Cannot access 'defaultExecResult$1' before initialization
    at refresh$3 (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:1889:5)
    at Module.<anonymous> (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:1804:1)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1175:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Module._load (node:internal/modules/cjs/loader:829:12)
    at c._load (node:electron/js2c/asar_bundle:5:13339)
    at loadApplicationPackage (/home/hellowodl/hyperplay/hyperplay-desktop-client/node_modules/electron/dist/resources/default_app.asar/main.js:121:16)
    at Object.<anonymous> (/home/hellowodl/hyperplay/hyperplay-desktop-client/node_modules/electron/dist/resources/default_app.asar/main.js:233:9)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1175:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Module._load (node:internal/modules/cjs/loader:829:12)
    at c._load (node:electron/js2c/asar_bundle:5:13339)
    at Object.<anonymous> (node:electron/js2c/browser_init:189:3102)
    at ./lib/browser/init.ts (node:electron/js2c/browser_init:189:3306)
(Use `electron --trace-warnings ...` to show where the warning was created)
(node:47213) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
```

This error is present on Linux as well, although I haven't found any adverse effects due to the error.

File: src/backend/main.ts (hunk 1 above)
Comment (BrettCleary) [reply src/backend/main.ts]: good idea. updated now

**Comment by BrettCleary** [comment]
> ```
> (node:47213) UnhandledPromiseRejectionWarning: ReferenceError: Cannot access 'defaultExecResult$1' before initialization
>     at refresh$3 (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:1889:5)
>     at Module.<anonymous> (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:1804:1)
>     at Module._compile (node:internal/modules/cjs/loader:1120:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1175:10)
>     at Module.load (node:internal/modules/cjs/loader:988:32)
>     at Module._load (node:internal/modules/cjs/loader:829:12)
>     at c._load (node:electron/js2c/asar_bundle:5:13339)
>     at loadApplicationPackage (/home/hellowodl/hyperplay/hyperplay-desktop-client/node_modules/electron/dist/resources/default_app.asar/main.js:121:16)
>     at Object.<anonymous> (/home/hellowodl/hyperplay/hyperplay-desktop-client/node_modules/electron/dist/resources/default_app.asar/main.js:233:9)
>     at Module._compile (node:internal/modules/cjs/loader:1120:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1175:10)
>     at Module.load (node:internal/modules/cjs/loader:988:32)
>     at Module._load (node:internal/modules/cjs/loader:829:12)
>     at c._load (node:electron/js2c/asar_bundle:5:13339)
>     at Object.<anonymous> (node:electron/js2c/browser_init:189:3102)
>     at ./lib/browser/init.ts (node:electron/js2c/browser_init:189:3306)
> (Use `electron --trace-warnings ...` to show where the warning was created)
> (node:47213) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> ```
> 
> This error is present on Linux as well, although I haven't found any adverse effects due to the error. Even on a clean install

this is fixed now

**Comment by BrettCleary** [comment]
> ## Tested on macOS M1 Ventura 13.2:
> ## Current Install
> * Already logged in on Epic and GOG and with 6 HP games installed and 2 Sideloaded.
> 
> ### Epic Games
> * [x]  Login (alternative)
> * [x]  Logout
> * [x]  Install
> * [ ]  Import Games
> * [x]  Launch Native ( this one is quite weird because it is asking to select a file now but it should ask for a folder, then we select the game folder and the import should work fine.)
> * [x]  Launch Wine-crossover
> * [x]  Move Game (fails in the end with an error on the dev tools console: `index.tsx:69 Uncaught (in promise) Error: Error invoking remote method 'moveInstall': TypeError: Cannot read properties of undefined (reading 'library')` Probably it is not changing the correct library at the end when calls changeInstallPath)
> * [ ]  Change Install Path (It moves the files instead of changing the path, but should only change the default path on installed.json file): `` (node:57248) UnhandledPromiseRejectionWarning: TypeError: Cannot read properties of undefined (reading 'library') at changeGameInstallPath$2 (/Users/flavio/develop/hyperplay-desktop-client/build/electron/main.js:5558:8) at Module.moveInstall$3 [as moveInstall] (/Users/flavio/develop/hyperplay-desktop-client/build/electron/main.js:4820:9) (node:57248) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 47) ``
> * [x]  Uninstall
> 
> ### GOG
> * [x]  Login
> * [x]  Logout
> * [x]  Install
> * [ ]  Import (Not working for both Windows and Native macOS games. I think same issue as above)
> * [x]  Launch Native
> * [x]  Launch Wine-crossover
> * [ ]  Move Game (Move the files but does not refresh the path after finishing, needs to click on the refresh button)
> * [ ]  Change Install Path (not working same issues as Epic's)
> * [x]  Uninstall
> 
> @hellowodl you can get Rocket league on Epic games for free to test when needed 👍🏽 Also, do not forget to test on a clean install as well.

import, move, and changeGameInstall should be fixed now for gog and legendary

**Comment by hellowodl** [comment]
Awesome. I'm running the test for Linux now

**Comment by hellowodl** [comment]
## Tested on Arch Linux:

## Current Install
- Clean install, no data is present within the HP application.

### Epic Games

- [x] Login (alternative)
- [x] Logout 
- [x] Install
- [ ] Import Games
- |- Says it succeeded, though after the gamelist refreshes the game is still not installed, no logs
- [ ] ~~Launch Native~~ 
- [x] Launch Proton*
- |- Rocket League tries to launch, but exits during the process, it might be related to my specific system and not to the refactor.
- [x] Move Game
 |- Works, though if the previous folder is ~/game and the next one is ~/game2, it moves the folder into ~/game2/game
- [x] Change Install Path
- [x] Uninstall

### GOG

- [x] Login
- [x] Logout 
- [x] Install
- [ ] Import
- |- Says it successfully imported the game, though the game still appears as not-installed
- [ ] ~~Launch Native~~
- [x] Launch Proton
- |- Proton appears to be running in the console, but contrary to the previous release, no window for the wine program is created, see log 1
- [x] Move Game (Move the files but does not refresh the path after finishing, needs to click on the refresh button)
- [x] Change Install Path
- [x] Uninstall

Log 1:

```
86018:0317/172619.567439:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 1 times!
(17:26:21) INFO:    [Backend]:          Launching The Settlers® 4: Gold Edition (1207659254)
(17:26:21) INFO:    [Backend]:          Preventing display from sleep
(17:26:21) INFO:    [Backend]:          Checking if wine version exists: Proton - Proton-GE-Proton7-51
(17:26:21) INFO:    [Backend]:          Checking if wine version exists: Proton - Proton-GE-Proton7-51
(17:26:21) WARNING: [Backend]:          You are using Proton, this can lead to some bugs. Please do not open issues with bugs related to games
(17:26:21) INFO:    [Backend]:          Checking if wine version exists: Proton - Proton-GE-Proton7-51
(17:26:21) INFO:    [Backend]:          Checking if wine version exists: Proton - Proton-GE-Proton7-51
(17:26:21) DEBUG:   [Backend]:          Running Wine command: run wineboot --init
(17:26:23) INFO:    [Backend]:          Creating abort controller for 1207659254
(17:26:23) INFO:    [Gog]:              Launching The Settlers® 4: Gold Edition: LD_PRELOAD= STEAM_COMPAT_CLIENT_INSTALL_PATH=/home/hellowodl/.steam/steam STEAM_COMPAT_DATA_PATH="/home/hellowodl/Games/HyperPlay/Prefixes/The Settlers 4 Gold Edition" PROTON_NO_ESYNC=1 PROTON_NO_FSYNC=1 STEAM_COMPAT_APP_ID=0 SteamAppId=0 SteamGameId="hyperplay-Settlers 4 Gold" PROTON_LOG_DIR=/home/hellowodl /home/hellowodl/hyperplay/hyperplay-desktop-client/public/bin/linux/gogdl launch "/home/hellowodl/Games/HyperPlay/Settlers 4 Gold" 1207659254 --no-wine --wrapper "'/home/hellowodl/.config/hyperplay/tools/proton/Proton-GE-Proton7-51/proton' run" --platform windows
(17:26:27) INFO:    [Backend]:          Deleting abort controller for 1207659254
(17:26:27) INFO:    [Backend]:          Stopping Display Power Saver Blocker
[85980:0317/172627.024673:ERROR:browser_main_loop.cc(269)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
[85980:0317/172627.024713:ERROR:browser_main_loop.cc(269)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
[85980:0317/172627.024735:ERROR:browser_main_loop.cc(269)] Gtk: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
(17:26:27) INFO:    [Frontend]:         Refreshing Library
(17:26:27) INFO:    [Gog]:              Getting GOG library
(17:26:27) INFO:    [Frontend]:         Refreshing Library
(17:26:27) INFO:    [Gog]:              Getting GOG library
[86018:0317/172627.098007:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 2 times!
[86018:0317/172627.099968:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 3 times!
(17:26:27) INFO:    [Gog]:              Number of library pages: 1
(17:26:27) INFO:    [Gog]:              Number of library pages: 1
(17:26:28) INFO:    [Gog]:              Saved games data
(17:26:28) INFO:    [Gog]:              Saved games data
```

Logs 2, unsure during which part of the process this showed up:
```
(17:46:30) ERROR:   [Gog]:              Setup: Looks like you have found new setup instruction, please report it on our Discord or GitHub appName: 1207664593, action: savePath
(17:46:30) ERROR:   [Backend]:          TypeError: Cannot read properties of undefined (reading 'replaceAll')
    at setup (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:6591:12)
    at async prepareWineLaunch (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:6906:5)
    at async Module.launch$1 [as launch] (/home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:4402:9)
    at async /home/hellowodl/hyperplay/hyperplay-desktop-client/build/electron/main.js:25669:26
    at async node:electron/js2c/browser_init:193:55
```

**Pull request merged**